core/src/main/scala/org/bitcoins/core/wallet/utxo/UTXOSpendingInfo.scala

```diff
@@ -527,132 +847,229 @@ case class P2WPKHV0SpendingInfo(
     outPoint: TransactionOutPoint,
     amount: CurrencyUnit,
     scriptPubKey: P2WPKHWitnessSPKV0,
-    signer: Sign,
+    override val signer: Sign,
     hashType: HashType,
     scriptWitness: P2WPKHWitnessV0)
-    extends SegwitV0NativeUTXOSpendingInfo {
+    extends SegwitV0NativeUTXOSpendingInfoFull
+    with SegwitV0NativeUTXOSpendingInfoSingle {
   require(P2WPKHWitnessSPKV0(signer.publicKey) == scriptPubKey,
           "Signer has incorrect public key")
   require(scriptWitness.pubKey == signer.publicKey,
           "Witness has incorrect public key")
 
-  override def signers: Seq[Sign] = Vector(signer)
+  override val requiredSigs: Int = 1
+
   override def conditionalPath: ConditionalPath =
     ConditionalPath.NoConditionsLeft
 }
 
-case class P2WSHV0SpendingInfo(
+sealed trait P2WSHV0SpendingInfo extends SegwitV0NativeUTXOSpendingInfo {
+  require(
+    CryptoUtil
+      .sha256(scriptWitness.redeemScript.asmBytes) == scriptPubKey.scriptHash,
+    s"Witness has incorrect script, got $scriptWitness")
+
+  override def scriptPubKey: P2WSHWitnessSPKV0
+  override def scriptWitness: P2WSHWitnessV0
+
+  def nestedSpendingInfo: RawScriptUTXOSpendingInfo
+}
+
+case class P2WSHV0SpendingInfoSingle(
+    outPoint: TransactionOutPoint,
+    amount: CurrencyUnit,
+    scriptPubKey: P2WSHWitnessSPKV0,
+    signer: Sign,
+    hashType: HashType,
+    scriptWitness: P2WSHWitnessV0,
+    conditionalPath: ConditionalPath)
+    extends P2WSHV0SpendingInfo
+    with SegwitV0NativeUTXOSpendingInfoSingle {
+  override val nestedSpendingInfo: RawScriptUTXOSpendingInfoSingle = {
```

Comment:
Maybe in a later PR we can just pass in the `nestedSpendingInfo` as a parameter, along with the `scriptPubKey`, and `scriptWitness`? This seems to be a pattern here where we re-construct the `nestedSpendingInfo` inside of the case class.